server/api.go

```diff
@@ -0,0 +1,1024 @@
+	if err != nil {
+		p.writeAPIError(w, &APIErrorResponse{
+			ID:         "cannot grant badge",
+			Message:    err.Error(),
+			StatusCode: http.StatusInternalServerError,
+		})
+		return
+	}
+
+	if shouldNotify {
+		p.notifyGrant(badgesmodel.BadgeID(badgeID), userID, grantToUser, notifyHere, req.ChannelId, reason)
+	}
+
+	p.mm.Post.SendEphemeralPost(userID, &model.Post{
+		UserId:    p.BotUserID,
+		ChannelId: req.ChannelId,
+		Message:   fmt.Sprintf("Badge `%s` granted to @%s.", badge.Name, grantToUser.Username),
+	})
+
+	dialogOK(w)
+}
+
+func (p *Plugin) dialogCreateSubscription(w http.ResponseWriter, r *http.Request, userID string) {
+	req := model.SubmitDialogRequestFromJson(r.Body)
+	if req == nil {
+		dialogError(w, "could not get the dialog request", nil)
+		return
+	}
+
+	u, err := p.mm.User.Get(userID)
+	if err != nil {
+		dialogError(w, err.Error(), nil)
+		return
+	}
+
+	if !canCreateSubscription(*u, req.ChannelId) {
+		dialogError(w, "You cannot create a subscription", nil)
+		return
+	}
+
+	typeIDStr, errText, errors := getDialogSubmissionTextField(req, DialogFieldBadgeType)
+	if errors != nil {
+		dialogError(w, errText, errors)
+		return
+	}
+
+	typeID, err := strconv.Atoi(typeIDStr)
+	if err != nil {
+		dialogError(w, "Invalid field", map[string]string{DialogFieldBadgeType: "Invalid type"})
+		return
+	}
+
+	err = p.store.AddSubscription(badgesmodel.BadgeType(typeID), req.ChannelId)
+	if err != nil {
+		dialogError(w, err.Error(), nil)
+	}
+
+	p.mm.Post.SendEphemeralPost(userID, &model.Post{
+		UserId:    p.BotUserID,
+		ChannelId: req.ChannelId,
+		Message:   "Subscription added",
+	})
+
+	dialogOK(w)
+}
+
+func (p *Plugin) dialogDeleteSubscription(w http.ResponseWriter, r *http.Request, userID string) {
+	req := model.SubmitDialogRequestFromJson(r.Body)
+	if req == nil {
+		dialogError(w, "could not get the dialog request", nil)
+		return
+	}
+
+	u, err := p.mm.User.Get(userID)
+	if err != nil {
+		dialogError(w, err.Error(), nil)
+		return
+	}
+
+	if !canCreateSubscription(*u, req.ChannelId) {
+		dialogError(w, "You cannot delete a subscription", nil)
+		return
+	}
+
+	typeIDStr, errText, errors := getDialogSubmissionTextField(req, DialogFieldBadgeType)
+	if errors != nil {
+		dialogError(w, errText, errors)
+		return
+	}
+
+	typeID, err := strconv.Atoi(typeIDStr)
+	if err != nil {
+		dialogError(w, "Invalid field", map[string]string{DialogFieldBadgeType: "Invalid type"})
+		return
+	}
+
+	err = p.store.RemoveSubscriptions(badgesmodel.BadgeType(typeID), req.ChannelId)
+	if err != nil {
+		dialogError(w, err.Error(), nil)
+	}
+
+	p.mm.Post.SendEphemeralPost(userID, &model.Post{
+		UserId:    p.BotUserID,
+		ChannelId: req.ChannelId,
+		Message:   "Subscription removed",
+	})
+
+	dialogOK(w)
+}
+
+func getDialogSubmissionTextField(req *model.SubmitDialogRequest, fieldName string) (value string, errText string, errors map[string]string) {
```

Comment:
the `errors` map here looks weird, I guess it is because helps in other place pass the data to the response, but I would expect to do it at the right level, you are validating a field with the `fieldName` `X`, so... you already know the fieldName from the outside, so you can get a regular error, and build the map outside.

Looks like here we are violating the single responsability principal, this function is parsing the input and formating the error to get ingested by other function.

Reply:
I guess this was an evolution while trying to get different dialog field types. You are right that it is not needed right now, but I still prefer it like this, for two reasons:
- I feel the parent function remains cleaner
- Errors in golang are supposed to start with lowercase and no punctuation, and I want a user facing string (starting with uppercase). If I go the error path I either ignore the error text, or add a wrapper, which I see as a poorer UX.
- If I keep the map here, there are less occurrences of the `map[string]string{fieldName: ...}` part, which also avoids potential mistakes when setting the field name.

Reply:
I dissagree about cleaner from parent function, it is hidding certain details that belongs to that level, not to the underneath level, so... I think is not cleaner, actually is less clear because you have to go to the underneath function to understand exactly what is expected.

Abotut the poorer UX I don't think so, it depends on how you proccess the error itself, it is not necesary to return a plain text error, you can return a more structured error and make the error message UX friendly, while you are still using a regular golang error to propagate the error state.

For example you can create a struct that represent a `ValidationError` that includes the field name and the data neede to rebuild an UX friendly message, and that `ValidationError` will have a `Error` method that returns the "lower case not puntuation mark" value of it. And then in the upper function you can try to cast the error to `ValidationError` with `errors.As` and if the casting is right, you generate the right error for the UX.

Again, this is an improvement over what you have there, feel free to keep it as is now, but I think is cleaner in the other way.